File: climb_stairs_with_DP.py

# Add DP (Top-Bottom Approach)
class Solution:
    def climbStairs(self, n: int) -> int:
        def dp(i):
            if i <= 2: 
                return i
            if i not in memo:
                # Instead of just returning dp(i - 1) + dp(i - 2), calculate it once and then
                # store the result inside a hashmap to refer to in the future.
                memo[i] = dp(i - 1) + dp(i - 2)
            
            return memo[i]
        #Dictionary Type
        memo = {}
        return dp(n)

def main():
    wayofstairs_class = Solution()
    test_result = wayofstairs_class.climbStairs(3)
    print(f"kimc: {test_result}")

if __name__ == "__main__":
    main()

File: DataS_Algorithm_Sample/Dynamic_Program/climb_stairs_with_DP.c
#include <stdio.h>

int memo_wayofup[46] = {1,1,0,};

/* This is framework of DP(especially Top-Bottom - Memorization) */
int climbStairs(int n) {
    if (memo_wayofup[n] != 0)
        return memo_wayofup[n];
    
    memo_wayofup[n] = climbStairs(n-1) + climbStairs(n-2);

    return memo_wayofup[n];
}

int main ()
{
    int input_stairs = 3;

    printf("Result : %d \n", climbStairs(input_stairs));

    return 0;
}